plugins/auth-backend/src/providers/factories.ts

```diff
@@ -48,3 +48,14 @@ export function createAuthProvider(
 
   return factory(options);
 }
+
+export function registerAuthProvider(
+  providerId: string,
+  providerFactory: AuthProviderFactory
+) {
+  if(factories[providerId]) {
+    throw Error(`Factory function for ${providerId} already exists`);
+  } else {
+    factories[providerId] = providerFactory;
```

Comment:
Would rather pass these to `createRouter` instead of accessing global state

Reply:
What would you recommend I pass to `createRouter`  ? Would it be a map of `providerId` to  `AuthProviderFactory`  , so that, when the `router` is reading the list of `auth.providers` , it looks into the map provided as parameter for a `providerId`, before checking the global `factory` (via `createAuthProvider`) ?

Reply:
Yeah it's prolly the thing that makes the most sense, thinking you'd be able to do something like this in the backend when setting up the plugin:

```ts
createRouter({
  ...otherThings,
  providers: {
    ...defaultProviderFactories,
    myprovider: myCustomProviderFactory,
  }
})
```

We'd export the current factories as `defaultProviderFactories` so that you can include or omit those as you want. That way you can replace existing ones in a pretty clear way too.

Was thinking about whether we should be working with providers directly in the app instead, but that prolly doesn't make sense as you'd kinda have to extract most of the `createRouter` implementation into a separate thing to recreate what you get in `AuthProviderFactoryOptions`

Reply:
The `createRouter` is quite convenient as it is one place where all the providers are initialized.. Most importantly, I think implementing the `tokenIssuer` is a lot of work and it takes care of that .
But so far the feedback is great ! I think i have  a good idea of the changes we'd need and will try to provide a clean way of setting up custom providers.

Reply:
Realizing that there's a piece that we need to take care of as well, which is that we likely need to start passing in the `providerId` as a parameter to the provider factories. Should be internal to the auth plugin, but we can't have them hardcoded within the provider factory function anymore when users of the plugin may be providing their own.

Reply:
Can I address this in the same PR as well ?

Reply:
Yep it's best if we do I think. Without actually having looked at the code and verified this, I think it should be as simple as adding `providerId: string` as part of the factory options passed to each factory, and then pass that ID in where the instantiation happens.

Reply:
Included this in the PR as well. Please have a look.